### PR TITLE
Update how FSDP uses state dicts for Torch 2.0

### PR DESCRIPTION
**Pull request by VikParuchuri**
Torch 2.0 has made some changes with FSDP and state dicts:

- `full_optim_state_dict` is [deprecated](https://pytorch.org/docs/stable/_modules/torch/distributed/fsdp/fully_sharded_data_parallel.html#FullyShardedDataParallel.full_optim_state_dict), and should be replaced with `optim_state_dict`.
- `scatter_full_optim_state_dict` is [deprecated](https://pytorch.org/docs/stable/_modules/torch/distributed/fsdp/fully_sharded_data_parallel.html#FullyShardedDataParallel.scatter_full_optim_state_dict), and should be replaced with `optim_state_dict_to_load`
- `optim_state_dict` does [not support](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.optim_state_dict) `optim_input`
- `set_state_dict_type` now takes an [optimizer state dict](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.set_state_dict_type), along with a model state dict

I have updated the FSDP plugin to mirror these changes.  I used `set_state_dict` to request the full state dict as needed.  I tried to keep parity with the existing logic in `save_model`, and only apply these changes for new versions of torch.

This works fine in my limited testing on Torch 2.0.

### CODE REVIEW

**Comment by VikParuchuri** [comment]
Note, this will not work until this PyTorch bugfix PR is merged - https://github.com/pytorch/pytorch/pull/97110

**Comment by sgugger** [comment]
cc @pacman100

File: src/accelerate/utils/dataclasses.py [hunk 1]
```diff
@@ -861,7 +872,7 @@ def save_model(self, accelerator, model, output_dir, model_index=0):
             with FSDP.state_dict_type(model, self.state_dict_type, self.state_dict_config):
                 state_dict = model.state_dict()
         else:
-            FSDP.set_state_dict_type(model, self.state_dict_type, self.state_dict_config)
+            FSDP.set_state_dict_type(model, self.state_dict_type, self.state_dict_config, self.optim_state_dict_config)
```
Comment (pacman100) [inline src/accelerate/utils/dataclasses.py:875]: is `self.optim_state_dict_config` required for saving model. If yes, I assume the same param should be passed in `load_model`, right?

**Review by pacman100 — COMMENTED** [review]
Thank you @VikParuchuri for all the fixes and pointers 🤗! Left a comment

File: src/accelerate/utils/dataclasses.py (hunk 1 above)
Comment (VikParuchuri) [reply src/accelerate/utils/dataclasses.py:875]: From looking at the PyTorch code, the optimizer config shouldn't be needed in either load or save model.  I'll remove it from here.

**Comment by VikParuchuri** [comment]
@pacman100 The PyTorch PR has now been merged.  Let me know how you want me to handle versioning, and I can finalize this.

**Comment by sgugger** [comment]
@pacman100 friendly ping

**Comment by VikParuchuri** [comment]
That's fine by me!

**Comment by pacman100** [comment]
Hello @VikParuchuri, 2.0.1 version of PyTorch is released. Please add that as the version checking, resolve the quality checks and then we are good to go. Thank you!

**Comment by pacman100** [comment]
Hello @VikParuchuri, Thank you for all the work. As we couldn't merge this PR then, clubbed these changes and fixed some issues with these in the above PR and added you as a co-author.

**Comment by pacman100** [comment]
Thank you!

**Pull request closed without merging**